pkg/controllers/elasticsearch/nodepool/util_test.go

```diff
@@ -0,0 +1,126 @@
+package nodepool
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+)
+
+func TestESImageToUse(t *testing.T) {
+	type testT struct {
+		name          string
+		spec          *v1alpha1.ElasticsearchClusterSpec
+		expectedImage *v1alpha1.ElasticsearchImage
+		expectedError bool
+	}
+	tests := []testT{
+		{
+			name: "version specified with no manual image spec",
+			spec: &v1alpha1.ElasticsearchClusterSpec{
+				Version: "6.1.1",
+			},
+			expectedImage: &v1alpha1.ElasticsearchImage{
+				FsGroup: 1000,
+				ImageSpec: v1alpha1.ImageSpec{
+					Repository: defaultElasticsearchImageRepository,
+					Tag:        "6.1.1",
+					PullPolicy: defaultElasticsearchImagePullPolicy,
+				},
+			},
+		},
+		{
+			name: "version specified with manual image spec",
+			spec: &v1alpha1.ElasticsearchClusterSpec{
+				Version: "6.1.1",
+				Image: &v1alpha1.ElasticsearchImage{
+					FsGroup: 1234,
+					ImageSpec: v1alpha1.ImageSpec{
+						Repository: "abcd",
+						Tag:        "xyz",
+						PullPolicy: "www",
+					},
+				},
+			},
+			expectedImage: &v1alpha1.ElasticsearchImage{
+				FsGroup: 1234,
+				ImageSpec: v1alpha1.ImageSpec{
+					Repository: "abcd",
+					Tag:        "xyz",
+					PullPolicy: "www",
+				},
+			},
+		},
+		{
+			name: "no version specified with manual image spec",
+			spec: &v1alpha1.ElasticsearchClusterSpec{
+				Version: "",
+				Image: &v1alpha1.ElasticsearchImage{
+					FsGroup: 1234,
+					ImageSpec: v1alpha1.ImageSpec{
+						Repository: "abcd",
+						Tag:        "xyz",
+						PullPolicy: "www",
+					},
+				},
+			},
+			expectedImage: &v1alpha1.ElasticsearchImage{
+				FsGroup: 1234,
+				ImageSpec: v1alpha1.ImageSpec{
+					Repository: "abcd",
+					Tag:        "xyz",
+					PullPolicy: "www",
+				},
+			},
+		},
+		{
+			name: "invalid (non semver) version specified and no manual image",
+			spec: &v1alpha1.ElasticsearchClusterSpec{
+				Version: "not.semver",
+			},
+			expectedError: true,
+		},
+		// We don't error here - if the specified version is non semver but the
+		// user has specified a manual image, validation should detect the
+		// invalid semver and fail.
```

Comment:
Shouldn't we always error with non-semver version?
Or does this comment mean that the non-semver  version will be caught and rejected by resource  validation in #199?

Reply:
Precisely that - yep. It seems odd to return an error in this case. The function should be returning the image that should be used for the spec, and the image is explicitly defined in the manifest, so there's not really an error.

There *is* however an error with the overall state/spec of the cluster, and validation should be responsible for detecting this. I guess more gates isn't a bad thing, but I think the gate might need to be moved to some other function if we do add it. I'm working on validation now and will expand that PR to encompass new fields added here once merged.